### ISSUE TITLE
LazyGetProperty() helper method

### ISSUE THREAD

**Issue by Chicagoan2016** [MarimerLLC/cslaforum#423]
I was adding a childlist to a parent object and I wanted to 'lazy load' the child collection, I noticed the snippets created a property with 'older' syntax which is  different than the current code base of project tracker sample.
The project tracker is using LazyGetProperty method. Is this the preferrable syntax now?

regards

**Comment by fujiiface** [comment]
@Chicagoan2016 I have a similar thread open right now: https://github.com/MarimerLLC/cslaforum/issues/324

The current LazyLoad syntax, as far I have been told and seen is as follows:
```
public static readonly PropertyInfo<CaseLetters> LettersProperty = RegisterProperty<CaseLetters>(c => c.Letters, "Letters", RelationshipTypes.Child | RelationshipTypes.LazyLoad);
[Display(Name = "Letters")]
public CaseLetters Letters
{
    get
    {
#if ASYNC
        return LazyGetPropertyAsync(LettersProperty, DataPortal.FetchAsync<CaseLetters>(ReadProperty(CaseIdProperty)));
#else
        return LazyGetProperty(LettersProperty, () => DataPortal.Fetch<CaseLetters>(ReadProperty(CaseIdProperty)));
#endif
    }
    private set { LoadProperty(LettersProperty, value); }
}
```

I want to create an MVC sample when I have time that will also show Insert/Update/Delete or update current examples but I haven't gotten around to it yet.

**Comment by tfreitasleal** [comment]
Good point. I thought I got it all right. I even made my own demo
https://github.com/tfreitasleal/CslaExtremeDemos/tree/master/LazyProperties
but it turns out I broke ProjectTracker. I'm investigating the issue but since I have projects that use `DataPortal.CreateChild` and others that use `DataPortal.CreateChild` and `DataPortal.Fetch` I'd say it depends on the child data portal.

For ProjectTracker XYZ is `CreateChild`
```
public ResourceAssignments Assignments
{
  get
  {
    return LazyGetProperty(AssignmentsProperty,
      () => DataPortal.XYZ<ResourceAssignments>(ReadProperty(IdProperty)));
  }
  private set { LoadProperty(AssignmentsProperty, value); }
}
```

**Comment by Chicagoan2016** [comment]
@tfreitasleal,
Thanks and please let us know of any changes

**Comment by tfreitasleal** [comment]
As I told you before, have a good look at my demo above, because the UI is truly lazy loading the child collection.

Let's elaborate a bit about LazyLoad.

1) First of all, the UI must be able to LazyLoad, say load a collection only when it shows a tab contents, or when you click a button to populate a list/grid. ProjectTracker doesn't work like that as it always shows all the object data.

2) At the BO level, the parent object must LazyLoad at all times, including when fetching data from the database and assign data to properties. There must be two different moments for fetch/assign: a **fetch/assign all but lazy loaded data** and a **fetch/assign lazy loaded data**. ProjectTracker fetches all data from the database at the same time and populates the child collection property all at the same time (in DataPortal_Fetch).

3) As I said in 2) above, the DAL must also know about LazyLoad. The DAL shouldn't loose time fetching the collection data when it loads the main object data, but only when you need to show the lazy loaded data. After all, the single purpose of lazy loading is to speed up the UI and delay the fecthing of heavy objects until they are absolutely needed..

In what concerns LazyLoad, ignore ProjectTracker and go for the demo sample referred a few messages earlier.

**Comment by tfreitasleal** [comment]
I'm afraid I can't help you. I suggest you have a look at CslaMVC sample.

**Comment by ajj7060** [comment]
Its been a while since I did anything advanced with MVC & Csla, but when I was I just created ModelBinder subclasses.  The model binder for the Root Csla object would check for an ID and either call the New or Get factory as needed.  Then delegated the actual mapping of posted values to the object's properties via the DefaultModelBinder.  I think I was subclassing DMB.

**Comment by fujiiface** [comment]
Based on the MVC ebook it looks like I need to create ChildObjectCreator and ChildObjectUpdater command objects since the stateless web interface is a different scenario from smart client applications.

**Comment by Chicagoan2016** [comment]
@fujiiface , I prefer writing one object similar to RoleEditManager.cs in ProjectTracker, just my two cents : )

**Comment by ajj7060** [comment]
@fujiiface That's another way to go.  I just found it easier to fetch the latest from the DB each time.

**Comment by fujiiface** [comment]
@Chicagoan2016 Briefly looking at RoleEditManager.cs, it looks like that is exactly what the ChildObjectCreator/Updaters do.  I just called them more generic than what they are in the book.

@ajj7060 I'm not sure what you mean exactly.  From past implementations we have usually used the child `Id` to get a BusinessBase object and insert/update it from directly then reload the parent object with the new list.  It's always felt like a hack to me though.  Is this what you're talking about?

**Comment by fujiiface** [comment]
@Chicagoan2016 My apologies.  I was looking at this thread again and think I misread your last response.  Both creator and updater in one object is the way I would go, too.  :)

**Comment by ajj7060** [comment]
@fujiiface No, I'd always just load the root object and the figure out which children needed to be inserted/updated/deleted, the save the root.

I think it sounds like you're talking about a switchable object (can be either root or child).  

Any of the approaches in this thread are fine I think of deal with MVC, there's just no great way of handling this.  I thought at one point the CslaModelBinder was going to be updated to handle children, maybe that ticket is still on the backlog.  

Still better than what I used to do on WebForms, which was keep the Csla object alive in the HttpCurrent.Session 😄

**Comment by fujiiface** [comment]
@ajj7060 Unfortunately I still have a few important (and complicated) pages keeping a Csla object alive in Session 😢
